Allow intercepting GLTFDocument on scene import

RedMser opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/4968

- Add virtual method `_initialize_gltf` to all glTF-based importers (gltf/glb, fbx, blend), which can intercept the step between creation and usage of `GLTFDocument` and `GLTFState`.
  - User would create a plugin which uses `add_scene_format_importer_plugin` to register a `EditorSceneFormatImporter`. It can inherit from e.g. the `GLTF` importer and merely override that specific virtual method to conditionally modify the document/state.
- With this, the user can e.g. create scripts that inherit from `GLTFDocumentExtension` and let those run for all imported scene assets that go through the glTF pipeline (filtering by path is possible). Among other things, this allows for direct access to the ["extensions"](https://github.com/KhronosGroup/glTF/blob/main/extensions/README.md#about-gltf-extensions) and ["extras"](https://kcoley.github.io/glTF/extensions/#extensions-vs-extras) dictionaries of the glTF nodes.
- MRP for testing: [gltf_document_extensions.zip](https://github.com/godotengine/godot/files/9438452/gltf_document_extensions.zip)
  - Re-importing the .gltf file should log a bunch of information, including the "extensions" section. Change the `extends` inside of `plugin.gd` to try out .fbx and .blend (don't forget to reload the plugin!).
  - The .blend file in specific has objects which get scaled based on a custom property variable.
  - The `godot-props-test.py` script can be installed and enabled as a Blender add-on to try out the user extension functionality, which will cause a glTF extension to be printed on import.

Example code:

```gdscript
# Create a plugin in project settings, then replace plugin.gd with this:
@tool
extends EditorPlugin

var importer

func _enter_tree():
	importer = MyImporter.new()
	add_scene_format_importer_plugin(importer, true)

func _exit_tree():
	remove_scene_format_importer_plugin(importer)

# Base class can be changed to FBX, Blend, etc.
class MyImporter extends EditorSceneFormatImporterGLTF:
	func _initialize_gltf(original_path, gltf_path, document, state):
		prints("orig path:", original_path, "// gltf path:", gltf_path)
		document.extensions.append(MyDocumentExtension.new())

class MyDocumentExtension extends GLTFDocumentExtension:
	func _import_node(state, gltf_node, json, node):
		prints("parent:", node.get_parent().name)
		prints("scene_name:", state.scene_name)
		prints("json:", json)
		return 0
```

RedMser committed:
Allow intercepting GLTFDocument on scene import

- Create EditorSceneImporterGLTFBase, which is how all scene importers
  that use glTF as backing format can share behavior.
- Allow user to intercept creation of GLTFDocument with virtual method
  _initialize_gltf, for example to conditionally add GLTFDocumentExtensions
- De-duplicate file extension check by moving it into the
  ResourceImporterScene class.

fire commented:
Is it possible to avoid the use of class name. I think the similar system of an array .gd files may work.

fire commented:
I have made a gltf extensions sandbox which was using the older approach.

https://github.com/V-Sekai/godot-gltf-sandbox I can port it to the newer approach.

fire commented on modules/gltf/editor/editor_scene_importer_fbx.cpp:
```diff
@@ -92,6 +93,18 @@ Node *EditorSceneFormatImporterFBX::import_scene(const String &p_path, uint32_t
+			gltf->get_extensions().append(document_extension);
+		} else if (r_err) {
+			*r_err = ERR_SCRIPT_FAILED;
+			ERR_PRINT("Invalid glTF document extension resource was specified.");
```
Can you add the classname of the gltf document extension?

fire commented on modules/gltf/editor/editor_scene_importer_gltf.cpp:
```diff
@@ -49,18 +50,44 @@ void EditorSceneFormatImporterGLTF::get_extensions(List<String> *r_extensions) c
+		return;
+	}
+
+	r_options->push_back(ResourceImporter::ImportOption(PropertyInfo(Variant::ARRAY, "import_script/extensions", PROPERTY_HINT_ARRAY_TYPE, vformat("%s/%s:%s", Variant::OBJECT, PROPERTY_HINT_RESOURCE_TYPE, "GLTFDocumentExtension")), Array()));
```
I believe this can be done better. 

The current flaw with this is that we to type repeatedly per new gltf extended class.

```c++
r_options->push_back(ResourceImporter::ImportOption(PropertyInfo(Variant::ARRAY, "import_script/extensions", PROPERTY_HINT_ARRAY_TYPE, vformat("%s/%s:%s", Variant::OBJECT, PROPERTY_HINT_RESOURCE_TYPE, "GLTFDocumentExtension")), Array()));```

Any ideas how to remove this duplication?

fire replied:
Is it possible to push from EditorSceneFormatImporterFBX and EditorSceneFormatImporterBlend to the superclass?

RedMser replied:
> Any ideas how to remove this duplication?

I initially had it in `ResourceImporterScene::get_import_options` but I think not all scene formats will be using glTF behind the scenes (so having an array of `GLTFDocumentExtension` would not make sense for them). I haven't looked enough at the code to find out how it differs from `EditorSceneFormatImporter`, but I assume it would also be the same issue then?

I could make a superclass method like `add_document_extensions(p_document, p_extensions)` possibly, so it's a one-liner per importer?

> Is it possible to push this check ...

I assume you wanted to comment on the hardcoded `!= "gltf"` etc.? I can change the code to use the extensions list method yes.

RedMser commented:
> Is it possible to avoid the use of class name.

`RichTextEffect` is another system that does it the same way (by creating scripts that use `class_name`). Don't know what alternatives there really are. I can't think of other systems in Godot that use the inspector to assign custom functionality (besides Import Script I guess?)

RedMser commented:
@fire Did some bigger changes, which I think makes the code a bit clearer:

- Create new base class `EditorSceneFormatImporterGLTFBase` from which GLTF/FBX/Blend inherit. It contains all shared functionality between the glTF-dependent formats:
  - New import options that all formats use.
  - The creation and conversion of the GLTFDocument, including extensions.
- Moved the extension check from `EditorSceneFormatImporter::get_import_options` up, so the caller has to filter for it now (in this case `ResourceImporterScene::get_import_options`).

fire commented:
Did you see the style check failures?

fire commented:
>  add bind_method for generate_gltf? 

I think so. We can also do the thing where we wait for people to complain. Since if it was essential the compiler and tests would have failed.

RedMser commented:
Hmm on second thought, I think since `import_scene` and similar methods aren't bound either, I think it's okay to keep it like this. If people need it in GDScript, it needs a bigger change anyway, since the raw pointers are causing template errors when combined with `bind_method`...

fire commented:
I realized there's an alternative away to do the array.

https://github.com/godotengine/godot/blob/master/scene/3d/skeleton_3d.cpp#L167-L171


```C++
p_list->push_back(
			PropertyInfo(Variant::OBJECT, "import_script/extensions",
					PROPERTY_HINT_RESOURCE_TYPE,
					"GLTFDocumentExtension",
					PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_DEFERRED_SET_RESOURCE | PROPERTY_USAGE_DO_NOT_SHARE_ON_DUPLICATE));
```

fire commented:
Can revise for a future pr.

RedMser commented:
> I realized there's an alternative way to do the array.

> Can revise for a future pr.

I think it would cause another situation like #60072 if it was implemented the same way as the modification stacks. Maybe needs a better solution on core level, but I think it's fine as it is, if it's documented well (once #49524 landed, at least).

fire commented:
We are removing the modification stacks and changing them to be nodes. https://github.com/godotengine/godot-proposals/issues/4863

I agree, it's fine as it is.

reduz commented:
Why not register this globally? I am not entirely sure what is the benefit of having it per scene. If you want some cextra onfiguration per scene, you can add an importer plugin.

RedMser commented:
> Why not register this globally?

`GLTFDocumentExtension` does not have a way to know what original file it's coming from. Best we have is `state.scene_name` which does not include the path, and may include a md5 hash for generated files.

> I am not entirely sure what is the benefit of having it per scene.

For glTF extensions I agree, having those global would make a lot more sense. But the "extras" object is likely only relevant for specific assets only.

> If you want some extra configuration per scene, you can add an importer plugin.

Do you mean `EditorImportPlugin` or `EditorSceneFormatImporter`?

How should the user get access to `GLTFDocument` there *before* the import is done, without re-implementing e.g. the .blend import procedure by hand? I could imagine adding a virtual method to the new `EditorSceneFormatImporterGLTFBase` class for example, which allows hijacking the glTF initialization.

fire commented:
> GLTFDocumentExtension does not have a way to know what original file it's coming from.

The file can be a runtime glb loaded from a in-memory buffer. So there is no file!

RedMser commented:
I have updated the approach with reduz's feedback. I also updated the MRP for testing above.

- Instead of an import option, create a virtual method `EditorSceneFormatImporterGLTFBase::_initialize_gltf` which does nothing by default.
- User can override it in a custom importer to (conditionally) add document extensions or do other things.

This allows for both global import behavior changes, or to check file path beforehand.

Sample code:
```
@tool
extends EditorSceneFormatImporterGLTF

func _initialize_gltf(gltf_path, document, state):
	document.extensions.append(load(MyDocumentExtension.new())
```
A simple `EditorPlugin` that uses `add_scene_format_importer_plugin(MyPlugin.new(), true)` is needed to register this (the `true` is necessary, otherwise the default gltf importer is executed first!).

---

Just noticed that you can't override `EditorSceneFormatImporterBlend::_import_scene` in gdscript (it does not get run unless if you are overriding `EditorSceneFormatImporter` specifically). Any idea how to fix this?

RedMser commented:
Does not seem possible to make a method (`_import_scene`) overridable in gdscript, while also calling the C++ method...

So I added an argument to retrieve the original file path (e.g. .blend path) to `_initialize_gltf` so you can do conditional checks on it.

Means that for now you can't really do any post-processing logic after the scene import is done (besides using import scripts like before).

fire commented:
> A simple EditorPlugin that uses add_scene_format_importer_plugin(MyPlugin.new(), true) is needed to register this (the true is necessary, otherwise the default gltf importer is executed first!). 

Does this mean we have to make a plug per type of gltf and we'll can only make one of this per extension?

RedMser commented:
@fire You have to create one editor plugin (in the addons folder) that looks basically like this:

```
@tool
extends EditorPlugin

const MyImporter = preload("res://addons/...")
var importer

func _enter_tree():
	importer = MyImporter.new()
	add_scene_format_importer_plugin(importer, true)

func _exit_tree():
	remove_scene_format_importer_plugin(importer)
```

but you can of course register multiple importer plugins in one editor plugin. The MRP zip has a simple example of how it is done.

Each importer plugin you add can change its logic freely based on scene file path (so it can add none, one, or multiple GLTFDocumentExtensions).
But you'll need to create a new importer plugin per scene file format (so `extends EditorSceneFormatImporterGLTF`, `Blend`, `FBX` etc. have to be separate plugins as far as I understand it - maybe you can re-use them idk).
Think it's ok since one project usually has one asset pipeline anyway.

fire commented:
I am not sure what is blocking this. There's a rebase problem though.

RedMser commented:
From my perspective it should be ready now.

Fixed the conflict that came with https://github.com/godotengine/godot/commit/805ffdfbf6841ac19c6fdbf3425a6ff15115bed7 but still gotta test whether everything works.

RedMser commented:
@fire Is anything still needed to get this merged?

I updated the example project so the addon is a single .gd file now - see PR description for the code. It's a bit simpler than before, might be good for putting into a distant-future wiki page or something.

fire commented:
Sorry, we was focusing on getting https://github.com/godotengine/godot/pull/62499 to be part of Godot 4 beta, and it is in a  limboed reviewing state by @reduz , I think both touch the same areas, and I want both.

The maintainers have been working hard, https://github.com/godotengine/godot/pulse/monthly and I've been trying to get the two reviews required.

1. The first review is another maintainer.
2. The second review from @akien-mga or others for style and architecture.

akien-mga reviewed (approved):
Style wise it seems fine to me, and the refactoring seems good. Still need an OK from @reduz.

fire commented:
@aaronfranke Please see if this is usable

aaronfranke commented:
The design of this seems quite limited, because it requires you to create a class that extends the importer in order to add extensions. This means that you are limited to using one importer at a time that adds all of the extensions. This will not work well if you have multiple third-party extensions, you would need to build a master class importer that includes the document extensions. It would make more sense to me to allow independently registering document extensions:

```gdscript
# Create a plugin in project settings, then replace plugin.gd with this:
@tool
extends EditorPlugin

func _enter_tree():
	# Extensions can be provided anywhere without having to know about each other or
	# a master scene importer, so multiple extensions will not conflict with each other.
	register_gltf_document_extension(MyDocumentExtension)

func _exit_tree():
	remove_scene_format_importer_plugin(MyDocumentExtension)

class MyDocumentExtension extends GLTFDocumentExtension:
	func _import_node(state, gltf_node, json, node):
		prints("parent:", node.get_parent().name)
		prints("scene_name:", state.scene_name)
		prints("json:", json)
		return 0
```



```gdscript
# Create a plugin in project settings, then replace plugin.gd with this:
@tool
extends EditorPlugin

func _enter_tree():
	# Extensions can be provided anywhere without having to know about each other or
	# a master scene importer, so multiple extensions will not conflict with each other.
	register_gltf_document_extension(OtherDocumentExtension)

func _exit_tree():
	remove_scene_format_importer_plugin(OtherDocumentExtension)

class OtherDocumentExtension extends GLTFDocumentExtension:
	func _import_node(state, gltf_node, json, node):
		prints("parent:", node.get_parent().name)
		prints("scene_name:", state.scene_name)
		prints("json:", json)
		return 0
```

RedMser commented:
@aaronfranke But then the use case for handling cases file-by-file is lost, since GLTFDocumentExtension does not have a file path association (iirc GLTFState only has the "generated" GLTF path, not the original file's). This is very important for handling the "extras" section (e.g. .blend file Custom Properties), which unlike "extensions" usually differs between the different assets.

aaronfranke commented:
@RedMser That's a good point, although that's not my use case. I think maybe we should have both systems in place, a way to register GLTFDocumentExtensions for GLTF importers to use, and a way to specify custom importers per-file. In that case, I don't really know what to think of with your PR, because I simply don't need it, the use case I have is to specify an extension that is used for importing all GLTF files.

fire commented:
https://github.com/godotengine/godot/pull/66026 fixes I and @aaronfranke' use cases.

akien-mga commented:
Superseded by #66026.

fire commented:
Let us know if you're able to solve your usecases as an mod to https://github.com/godotengine/godot/pull/66026. I am still interested.

RedMser commented:
@fire Thanks, I checked the PR and it seems to be a good alternative. Some things I have noticed:

- I tried using it from an EditorPlugin and noticed you can't unregister gltfdocumentextensions. This is a problem when disabling and re-enabling the plugin. I added an unregister method in #69022
- `GLTFState.base_path` only includes the **folder** path of the .gltf file, not the actual file name (so for example `res://` is all I get when importing `res://test.gltf`). Is this a bug or an intentional change? I don't remember this happening before, but I can open an issue if wanted.

I will close my proposal since this seems to be a good alternative.

fire commented:
Base path is used for the engine importer to find .bin and .png/.jpg. Like in the blender importer the gltf base path is in the .godot/importer/ folder.

I’ll poke @aaronfranke about unregistering gltf document extensions.
